### DISCUSSION TITLE
Error testing endpoints on Gitlab CI/CD

### DISCUSSION THREAD

**Discussion by kashyapm94** [fastapi/fastapi#6888] [Questions]
I have an app built on `fastapi` and for the routes, I am writing tests using `pytest` using the structure described [here](https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/tests)

My project structure is as follows:

    .
    ├── app
    │   ├── api
    │   │   ├── __init__.py
    │   │   └── v1
    │   │       ├── __init__.py
    │   │       ├── invest.py
    │   │       └── tokens.py
    │   ├── core
    │   │   ├── config.py
    │   │   ├── decode_verify_jwt.py
    │   │   ├── __init__.py
    │   │   └── token_checker.py
    │   ├── crud
    │   │   └── __init__.py
    │   ├── database
    │   │   └── __init__.py
    │   ├── __init__.py
    │   ├── main.py
    │   ├── models
    │   │   ├── __init__.py
    │   │   └── invest_input.py
    │   ├── schemas
    │   │   ├── __init__.py
    │   │   └── token.py
    │   ├── services
    │   │   ├── __init__.py
    │   │   └── invest_funcs.py
    │   └── tests
    │       ├── api
    │       │   ├── api_v1
    │       │   │   ├── __init__.py
    │       │   │   ├── test_invest.py
    │       │   │   └── test_login.py
    │       │   └── __init__.py
    │       ├── conftest.py
    │       ├── __init__.py
    │       └── utils
    │           ├── __init__.py
    │           └── utils.py
    ├── db_pv
    ├── package.json
    ├── package-lock.json
    ├── Pipfile
    ├── Pipfile.lock
    ├── README.md
    └── serverless.yml

In the `tokens.py` (this connects to a lambda hosted on AWS and gets the `id_token` back), I have the following:

    import json
    
    import boto3
    from fastapi import APIRouter, HTTPException, Depends
    from fastapi.security import OAuth2PasswordRequestForm
    
    from app.core.config import settings
    from ...schemas.token import Token
    
    router = APIRouter(
        tags=["token"]
    )
    lambda_client = boto3.client('lambda')
    
    
    @router.post("/token", response_model=Token)
    def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
        username = form_data.username
        password = form_data.password
        lambda_payload = {
            "username": username,
            "password": password
        }
        lambda_payload = json.dumps(lambda_payload)
        try:
            r = lambda_client.invoke(
                FunctionName=settings.TOKEN_GENERATOR_LAMBDA,
                InvocationType='RequestResponse',
                Payload=lambda_payload
            )
            read_r = json.loads(r['Payload'].read())
            return {'access_token': read_r['data']['id_token']}
        except Exception:
            raise HTTPException(
                status_code=400,
                detail="Invalid authentication credentials",
                headers={"WWW-Authenticate": "Bearer"},
            )

and the code in `invest.py`:

    import os
    
    import numpy as np
    from fastapi import APIRouter, Depends
    from fastapi.security import OAuth2PasswordBearer
    
    from ...models.invest_input import InvestModel
    
    router = APIRouter(
        prefix="/invest",
        tags=["invest"],
        responses={403: {'description': 'Forbidden'}}
    )
    oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")
    
    
    @router.get('/calc_invest', dependencies=[Depends(oauth2_scheme)])
    def get_invest_calc(params: InvestModel = Depends()):
        initial_invest = params.initial_invest
        interest_rate = params.interest_rate
        reinvest_rate = params.reinvest_rate
        return initial_invest, interest_rate, reinvest_rate

Both these routes work fine. I tried witing tests for them as follows:

`utils.py` contains

```
from typing import Dict

from fastapi.testclient import TestClient

from ...core.config import settings


def get_superuser_token_headers(client: TestClient) -> Dict[str, str]:
    login_data = {
        "username": settings.TEST_USER,
        "password": settings.TEST_USER_PASSWORD,
    }
    r = client.post("/token", data=login_data)
    tokens = r.json()
    id_token = tokens["access_token"]
    headers = {"Authorization": f"Bearer {id_token}"}
    return headers
```

`conftest.py` contents:

    from typing import Dict, Generator
    
    import pytest
    from app.main import app
    from fastapi.testclient import TestClient
    
    from .utils.utils import get_superuser_token_headers
    
    
    @pytest.fixture(scope="module")
    def client() -> Generator:
        with TestClient(app) as c:
            yield c
    
    
    @pytest.fixture(scope="module")
    def superuser_token_headers(client: TestClient) -> Dict[str, str]:
        return get_superuser_token_headers(client)

and the tests for the routes:

`test_login.py` is:

    from fastapi.testclient import TestClient
    
    from ....core.config import settings
    
    
    def test_get_access_token(client: TestClient) -> None:
        login_data = {
            "username": settings.TEST_USER,
            "password": settings.TEST_USER_PASSWORD,
        }
        response = client.post("/token", data=login_data)
        tokens = response.json()
        assert response.status_code == 200
        assert "access_token" in tokens
        assert tokens["access_token"]
    
    
    def test_wrong_access_token(client: TestClient) -> None:
        login_data = {
            "username": 'jack',
            "password": 'qwerty',
        }
        response = client.post("/token", data=login_data)
        assert response.status_code == 400
        assert response.json() == {"detail": "Invalid authentication credentials"}

and `test_invest.py`

    from fastapi.testclient import TestClient
    
    
    def test_calc_invest(
        client: TestClient,
        superuser_token_headers: dict
    ) -> None:
        data = {
            'initial_invest': 10000,
            'interest_rate': 0.05,
            'reinvest_rate': 0.01
        }
        response = client.get(
            '/invest/calc_invest',
            headers=superuser_token_headers,
            params=data
        )
        assert response.status_code == 200

**When I run the test on my local system by doing `pytest`, it works absolutely fine, but when I run the same using `Gitlab CI/CD`, I start getting the following error**:

    ============================= test session starts ==============================
    platform linux -- Python 3.7.10, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
    rootdir: /builds/EC/di/invest-tool
    plugins: cov-2.11.1
    collected 3 items
    app/tests/api/api_v1/test_invest.py E                                    [ 33%]
    app/tests/api/api_v1/test_login.py F.                                    [100%]
    ==================================== ERRORS ====================================
    ______________________ ERROR at setup of test_calc_invest ______________________
    client = <starlette.testclient.TestClient object at 0x7f840e5f4890>
        @pytest.fixture(scope="module")
        def superuser_token_headers(client: TestClient) -> Dict[str, str]:
    >       return get_superuser_token_headers(client)
    app/tests/conftest.py:18: 
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    client = <starlette.testclient.TestClient object at 0x7f840e5f4890>
        def get_superuser_token_headers(client: TestClient) -> Dict[str, str]:
            login_data = {
                "username": settings.TEST_USER,
                "password": settings.TEST_USER_PASSWORD,
            }
            r = client.post("/token", data=login_data)
            tokens = r.json()
    >       id_token = tokens["access_token"]
    E       KeyError: 'access_token'
    app/tests/utils/utils.py:15: KeyError
    =================================== FAILURES ===================================
    ____________________________ test_get_access_token _____________________________
    client = <starlette.testclient.TestClient object at 0x7f840e5daad0>
        def test_get_access_token(client: TestClient) -> None:
            login_data = {
                "username": settings.TEST_USER,
                "password": settings.TEST_USER_PASSWORD,
            }
            response = client.post("/token", data=login_data)
            tokens = response.json()
    >       assert response.status_code == 200
    E       assert 400 == 200
    E        +  where 400 = <Response [400]>.status_code
    app/tests/api/api_v1/test_login.py:13: AssertionError

I think that the `client.post("/token", data=login_data)` is causing the problem.

Is there a way I can solve this issue?

Thanks.

P.S: If it matters, my `config.py`:

```
from pydantic import BaseSettings


class Settings(BaseSettings):
    SQLALCHEMY_DATABASE_URI: str
    COGNITO_REGION: str
    COGNITO_USERPOOL_ID: str
    COGNITO_APP_CLIENT_ID: str
    TOKEN_GENERATING_LAMBDA: str
    PROJECT_NAME: str = "Invest Tool"
    BACKEND_CORS_ORIGINS: list = ["*"]
    TEST_USER: str = 'username'
    TEST_USER_PASSWORD: str = 'password'

    class Config:
        env_file = ".env"


settings = Settings()

```

**Comment by coder3112** [comment]
> I think that the  ```client.post("/token", data=login_data) ``` is  causing the problem

Actually ```client.post("/token", data=login_data) ``` is what sends the request
On line 15 ``` KeyError``` is there according to pytest in utils.py
 The problem is because there is no access_token key in the dictionary.

Maybe because there is no such credential in the database?

**Comment by kashyapm94** [comment]
@coder3112 I checked and rechecked everything.  I also hardcoded the username and password for testing it, and still the same error

**Comment by coder3112** [comment]
Could you please tell the value of ```r.json() ``` and ```r.status_code``` (the response)

**Comment by kashyapm94** [comment]
value of `r.json()` is `{'detail': 'Invalid authentication credentials'}` (even though the correct credentials are being used)

value of `r.status_code` is `400`.

Both these are being raised in the `tokens.py` script (code has been added above)

**Comment by valentin994** [comment]
How do you run the ci/cd test, dockers in the pipeline or? At what address are they running?  I haven't worked with aws as a provider but maybe your URL that you specify(if you do that) isn't targeting the right address as you are using gitlab runners for your instances?

**Comment by kashyapm94** [comment]
@valentin994 I run it in the pipeline directly and since I am testing it using `client.post`, it runs on `http://testserver/token`

**Comment by tiangolo** [comment]
Thanks for the help here @coder3112 and @valentin994 ! 👏  🙇 

I assume you solved it @kashyapm94, thanks for closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
